dissect/target/plugins/apps/browser/firefox.py

```diff
@@ -265,4 +295,324 @@ def downloads(self) -> Iterator[BrowserDownloadRecord]:
+    key = pbkdf2_hmac("sha256", k, entry_salt, iteration_count, dklen=key_length)
+
+    iv = b"\x04\x0e" + decoded_item[0][1][1][1].native
+    cipher_text = decoded_item[1].native
+    return AES.new(key, AES.MODE_CBC, iv).decrypt(cipher_text)
+
+
+def decrypt_sha1_triple_des_cbc(decoded_item: core.Sequence, primary_password: bytes, global_salt: bytes) -> bytes:
+    """
+    SEQUENCE {
+        SEQUENCE {
+            OBJECTIDENTIFIER 1.2.840.113549.1.12.5.1.3
+            SEQUENCE {
+                OCTETSTRING entry_salt
+                INTEGER 01
+            }
+        }
+        OCTETSTRING encrypted
+    }
+    """
+    entry_salt = decoded_item[0][1][0].native
+    cipher_text = decoded_item[1].native
+    key = decrypt_moz_3des(global_salt, primary_password, entry_salt, cipher_text)
+    return key[:24]
+
+
+def decrypt_master_key(decoded_item: core.Sequence, primary_password: bytes, global_salt: bytes) -> tuple[bytes, str]:
+    """At this stage, we're not yet sure of the structure of decoded_item.
+
+    The structure will depend on the object identifier at [0][0], hence we extract it.
+
+    SEQUENCE {
+        SEQUENCE {
+            OBJECTIDENTIFIER ???
+            ...
+        }
+        ...
+    }
+    """
+    if not HAS_CRYPTO:
+        raise ValueError("Missing cryptography dependency")
+
+    object_identifier = decoded_item[0][0]
+    algorithm = object_identifier.dotted
+
+    if algos.EncryptionAlgorithmId.map(algorithm) == "pbes2":
+        return decrypt_pbes2(decoded_item, primary_password, global_salt), algorithm
+    elif algorithm == pbeWithSha1AndTripleDES_CBC:
+        return decrypt_sha1_triple_des_cbc(decoded_item, primary_password, global_salt), algorithm
+    else:
+        # Firefox supports other algorithms (i.e. Firefox before 2018), but decrypting these is not (yet) supported.
+        return b"", algorithm
+
+
+def query_global_salt(key4_file: TargetPath) -> tuple[str, str]:
+    with key4_file.open("rb") as fh:
+        db = sqlite3.SQLite3(fh)
+        for row in db.table("metadata").rows():
+            if row.get("id") == "password":
+                return row.get("item1", ""), row.get("item2", "")
+
+
+def query_master_key(key4_file: TargetPath) -> tuple[str, str]:
+    with key4_file.open("rb") as fh:
+        db = sqlite3.SQLite3(fh)
+        for row in db.table("nssPrivate").rows():
+            return row.get("a11", ""), row.get("a102", "")
+
+
+def retrieve_master_key(primary_password: bytes, key4_file: TargetPath) -> tuple[bytes, str]:
+    if not HAS_CRYPTO:
+        raise ValueError("Missing cryptography dependency")
+
+    global_salt, password_check = query_global_salt(key4_file)
+    decoded_password_check = core.load(password_check)
+    decrypted_password_check, algorithm = decrypt_master_key(decoded_password_check, primary_password, global_salt)
+
+    if not decrypted_password_check:
+        log.warning("Encountered unknown algorithm %s while decrypting master key.", algorithm)
```

Comment:
Maybe replace all logs in this function with raises and do the logging one level higher?